services/middleware/src/transactions.rs

```diff
@@ -0,0 +1,192 @@
+    NoService = 0,
+    /// The service instance targeted by the checked call is not active.
+    ServiceIsNotActive = 1,
+    /// The called service instance has an unexpected artifact.
+    ArtifactMismatch = 2,
+    /// The called service instance has an unsupported version.
+    VersionMismatch = 3,
+}
+
+/// Checked call to the service.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, BinaryValue)]
+pub struct CheckedCall {
+    /// Expected name of the artifact.
+    pub artifact_name: String,
+    /// Version requirement(s) on the artifact.
+    pub artifact_version: VersionReq,
+    /// The call contents.
+    pub inner: AnyTx,
+}
+
+impl ProtobufConvert for CheckedCall {
+    type ProtoStruct = proto::CheckedCall;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        let mut pb = Self::ProtoStruct::new();
+        pb.set_artifact_name(self.artifact_name.clone());
+        pb.set_artifact_version(self.artifact_version.to_string());
+        pb.set_inner(self.inner.to_pb());
+        pb
+    }
+
+    fn from_pb(mut pb: Self::ProtoStruct) -> Result<Self, failure::Error> {
+        let artifact_name = pb.take_artifact_name();
+        let artifact_version = pb.get_artifact_version().parse()?;
+        let inner = AnyTx::from_pb(pb.take_inner())?;
+        Ok(Self {
+            artifact_name,
+            artifact_version,
+            inner,
+        })
+    }
+}
+
+#[test]
+fn checked_call_in_json() {
+    use exonum::runtime::CallInfo;
+    use serde_json::json;
+
+    let mut checked_call = CheckedCall {
+        artifact_name: "test-artifact".to_string(),
+        artifact_version: "^1.0.0".parse().unwrap(),
+        inner: AnyTx {
+            call_info: CallInfo::new(100, 0),
+            arguments: vec![],
+        },
+    };
+    assert_eq!(
+        serde_json::to_value(&checked_call).unwrap(),
+        json!({
+            "artifact_name": "test-artifact",
+            "artifact_version": "^1.0.0",
+            "inner": checked_call.inner,
+        })
+    );
+
+    checked_call.artifact_version = ">=0.9, <2".parse().unwrap();
+    assert_eq!(
+        serde_json::to_value(&checked_call).unwrap(),
+        json!({
+            "artifact_name": "test-artifact",
+            "artifact_version": ">= 0.9, < 2",
+            "inner": checked_call.inner,
+        })
+    );
+}
+
+/// Transactions executed in a batch.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[derive(ProtobufConvert, BinaryValue)]
+#[protobuf_convert(source = "proto::Batch")]
+pub struct Batch {
+    /// Transactions included in the batch.
+    pub inner: Vec<AnyTx>,
+}
+
+impl Batch {
+    /// Creates an empty batch.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Appends a call into the batch.
+    pub fn with_call(mut self, call: AnyTx) -> Self {
+        self.inner.push(call);
+        self
+    }
+}
+
+/// Transactional interface of the utilities service.
+#[exonum_interface]
+pub trait UtilsInterface {
```

Comment:
The `UtilsInterface` sounds a little bit obscure. What do you think about `CallUtils` name for this interface?

Reply:
I've forgotten to rename it to `MiddlewareInterface` 🤦‍♂